src/EditorFeatures/TestUtilities/Threading/WpfTestRunner.cs

```diff
@@ -61,35 +59,31 @@ protected override Task<decimal> InvokeTestMethodAsync(ExceptionAggregator aggre
             var sta = StaTaskScheduler.DefaultSta;
             var task = Task.Factory.StartNew(async () =>
             {
-                Debug.Assert(sta.StaThread == Thread.CurrentThread);
+                using var _ = await SharedData.TestSerializationGate.DisposableWaitAsync(CancellationToken.None);
+
+                // Reset our flag ensuring that part of this test actually needs WpfFact
+                s_wpfFactRequirementReason = null;
 
-                using (await SharedData.TestSerializationGate.DisposableWaitAsync(CancellationToken.None))
+                if (_passedTests.TryGetValue(Test.DisplayName, out var info))
                 {
-                    try
+                    return info.Time;
+                }
+                else
+                {
+                    var tcs = new TaskCompletionSource<decimal>();
+                    sta.Post(() =>
                     {
+                        Debug.Assert(sta.StaThread == Thread.CurrentThread);
                         Debug.Assert(SynchronizationContext.Current is DispatcherSynchronizationContext);
 
-                        // Reset our flag ensuring that part of this test actually needs WpfFact
-                        s_wpfFactRequirementReason = null;
+                        // Just call back into the normal xUnit dispatch process now that we are on an STA Thread with no synchronization context.
+                        var invoker = new XunitTestInvoker(Test, MessageBus, TestClass, ConstructorArguments, TestMethod, TestMethodArguments, BeforeAfterAttributes, aggregator, CancellationTokenSource);
+                        tcs.SetResult(invoker.RunAsync().JoinUsingDispatcher(CancellationTokenSource.Token));
```

Comment:
What happens if anything throws? Do we need to also do a tcs.SetException or something, otherwise the task we return never completes?